src/main/java/net/neoforged/neoforge/flag/Flag.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.flag;
+
+import com.google.common.collect.Maps;
+import com.mojang.serialization.Codec;
+import io.netty.buffer.ByteBuf;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * Class describing a specific flag.
+ */
+public final class Flag {
+    public static final Codec<Flag> CODEC = ResourceLocation.CODEC.xmap(Flag::of, flag -> flag.identifier);
+    public static final StreamCodec<ByteBuf, Flag> STREAM_CODEC = ResourceLocation.STREAM_CODEC.map(Flag::of, flag -> flag.identifier);
+
+    private static final Map<ResourceLocation, Flag> FLAGS = Maps.newConcurrentMap();
+    private static final Collection<Flag> FLAGS_VIEW = Collections.unmodifiableCollection(FLAGS.values());
+
+    private final ResourceLocation identifier;
+
+    private Flag(ResourceLocation identifier) {
+        this.identifier = identifier;
+    }
+
+    /**
+     * @return The owning mod id/namespace for this {@link Flag flag}.
+     */
+    public String namespace() {
+        return identifier.getNamespace();
+    }
+
+    /**
+     * @return The unique identifier for this {@link Flag flag}.
+     * @apiNote This identifier does not need to be globally unique, only unique to the owning mod.
+     */
+    public String identifier() {
```

Comment:
Instead of having separate identifier/namespace methods, these should return the backing ResLoc, to enable access to the rest of the ResLoc API

Reply:
But would anyone ever need access to the rest of the RL Api from a flags identifier as context? I'm pretty sure the answer is no but if not I can easily update this to return the full RL

The only reason a RL is used here is for namespace prefixed identifier parsing, I could drop the RL and do all that myself in the codec/stream codecs. Just found it easier to reuse a RL for this though.